src/runtime/refiner.ts

```diff
@@ -8,142 +8,699 @@
+    static fromExpression(expr: RefinementExpression): Range {
+        if (expr.kind === 'binary-expression') {
+            const bexpr = expr as BinaryExpression;
+            if (bexpr.leftExpr.kind === 'field-name' && bexpr.rightExpr.kind === 'number') {
+                return Range.makeInitialGivenOp(bexpr.operator.op, (bexpr.rightExpr as NumberPrimitive).value);
             }
-            return Refiner.applyOperator(expr.operator, [curr]);
+            const left = Range.fromExpression(bexpr.leftExpr);
+            const right = Range.fromExpression(bexpr.rightExpr);
+            return Range.updateGivenOp(bexpr.operator.op, [left, right]);
+        } else if (expr.kind === 'unary-expression') {
+            const uexpr = expr as UnaryExpression;
+            const rg = Range.fromExpression(uexpr.expr);
+            return Range.updateGivenOp(uexpr.operator.op, [rg]);
         }
-        // TODO(ragdev): Update when true, false and string literals are supported
-        // in the refinement expression.
-        if (expr.kind === 'number-node') {
-            return expr.value;
-        } else if (expr.kind === 'field-name-node') {
-            if (data[expr.value] !== undefined) {
-                return data[expr.value];
+        throw new Error(`Cannot resolve primitive nodes by themselves.`);
+    }
+
+    static unionOf(range1: Range, range2: Range): Range {
+        const newRange = Range.copyOf(range1);
+        newRange.union(range2);
+        return newRange;
+    }
+
+    static intersectionOf(range1: Range, range2: Range): Range {
+        const newRange = Range.copyOf(range1);
+        newRange.intersect(range2);
+        return newRange;
+    }
+
+    static complementOf(range: Range, from: Range = Range.infiniteRange()) {
+        return Range.difference(from, range);
+    }
+
+    // difference(A,B) = A\B = A - B
+    static difference(range1: Range, range2: Range): Range {
+        const newRange = new Range();
+        for (const seg of range1.segments) {
+            const ntrsct = Range.copyOf(range2);
+            ntrsct.intersectWithSeg(seg);
+            let from: Boundary = {...seg.from};
+            for (const iseg of ntrsct.segments) {
+                const to: Boundary = {...iseg.from};
+                to.kind = to.kind === 'open' ? 'closed' : 'open';
+                if (Segment.isValid(from, to)) {
+                    newRange.segments.push(new Segment(from, to));
+                }
+                from = iseg.to;
+                from.kind = from.kind === 'open' ? 'closed' : 'open';
+            }
+            const to: Boundary = {...seg.to};
+            if (Segment.isValid(from, to)) {
+                newRange.segments.push(new Segment(from, to));
+            }
+        }
+        return newRange;
+    }
+
+    equals(range: Range): boolean {
+        return JSON.stringify(this) === JSON.stringify(range);
```

Comment:
This assumes the segments are expressed in a consistent order; i.e. R(0..5, 20..30) won't equal R(20..30, 0..5). This will almost certainly bite us at some point.

Also, is it possible to have denormalised ranges?  Technically, R(0..20, 21..30) == R(0..20, 10..30) == R(0..30)

Reply:
Right. Is there any inbuilt way to compare two objects? Alternatively I could be explicit with my function and iterate over all the segments and compare them individually.

I don't believe R(0..20, 21..30) == R(0..20, 10..30) == R(0..30) is true. Specifically the first range does not contain values between 20 and 21.

But to answer the question about denormalised ranges, unless someone explicitly mutates the segments array of a range, as per the current behavior we cannot have denormalised ranges (by intentional design). I think I'll add some guards to prevent someone from inputting an "invalid" segment list.